hbase-server/src/main/java/org/apache/hadoop/hbase/procedure2/store/region/RegionProcedureStore.java

```diff
@@ -0,0 +1,576 @@
+    if (!fs.exists(replayEditsDir) && !fs.mkdirs(replayEditsDir)) {
+      throw new IOException("Failed to create replay directory: " + replayEditsDir);
+    }
+    Path walsDir = new Path(rootDir, HREGION_LOGDIR_NAME);
+    for (FileStatus walDir : fs.listStatus(walsDir)) {
+      if (!walDir.isDirectory()) {
+        continue;
+      }
+      if (walDir.getPath().getName().startsWith(factoryId)) {
+        LOG.warn("This should not happen in real production as we have not created our WAL " +
+          "directory yet, ignore if you are running a procedure related UT");
+      }
+      Path deadWALDir;
+      if (!walDir.getPath().getName().endsWith(DEAD_WAL_DIR_SUFFIX)) {
+        deadWALDir =
+          new Path(walDir.getPath().getParent(), walDir.getPath().getName() + DEAD_WAL_DIR_SUFFIX);
+        if (!fs.rename(walDir.getPath(), deadWALDir)) {
+          throw new IOException("Can not rename " + walDir + " to " + deadWALDir +
+            " when recovering lease of proc store");
+        }
+        LOG.info("Renamed {} to {} as it is dead", walDir.getPath(), deadWALDir);
+      } else {
+        deadWALDir = walDir.getPath();
+        LOG.info("{} is already marked as dead", deadWALDir);
+      }
+      for (FileStatus walFile : fs.listStatus(deadWALDir)) {
+        Path replayEditsFile = new Path(replayEditsDir, walFile.getPath().getName());
+        leaseRecovery.recoverFileLease(fs, walFile.getPath());
+        if (!fs.rename(walFile.getPath(), replayEditsFile)) {
+          throw new IOException("Can not rename " + walFile.getPath() + " to " + replayEditsFile +
+            " when recovering lease of proc store");
+        }
+        LOG.info("Renamed {} to {}", walFile.getPath(), replayEditsFile);
+      }
+      LOG.info("Delete empty proc wal dir {}", deadWALDir);
+      fs.delete(deadWALDir, true);
+    }
+    RegionInfo regionInfo = HRegionFileSystem.loadRegionInfoFileContent(fs, regionDir);
+    WAL wal = createWAL(fs, rootDir, regionInfo);
+    conf.set(HRegion.RECOVERED_EDITS_DIR,
+      replayEditsDir.makeQualified(fs.getUri(), fs.getWorkingDirectory()).toString());
+    return HRegion.openHRegionFromTableDir(conf, fs, tableDir, regionInfo, TABLE_DESC, wal, null,
+      null);
+  }
+
+  @SuppressWarnings("deprecation")
+  private void tryMigrate(FileSystem fs) throws IOException {
+    Configuration conf = server.getConfiguration();
+    Path procWALDir =
+      new Path(CommonFSUtils.getWALRootDir(conf), WALProcedureStore.MASTER_PROCEDURE_LOGDIR);
+    if (!fs.exists(procWALDir)) {
+      return;
+    }
+    LOG.info("The old procedure wal directory {} exists, start migrating", procWALDir);
+    WALProcedureStore store = new WALProcedureStore(conf, leaseRecovery);
+    store.start(numThreads);
+    store.recoverLease();
+    MutableLong maxProcIdSet = new MutableLong(-1);
+    MutableLong maxProcIdFromProcs = new MutableLong(-1);
+    store.load(new ProcedureLoader() {
```

Comment:
It could. If this happens the master will crash, i.e, we fail startup. We first need to fix the original wal procedure store, and then try to upgrade.